drivers/SmartThings/zigbee-window-treatment/src/aqara/init.lua

```diff
@@ -0,0 +1,258 @@
+      if device.preferences[opencloseDirectionPreferenceId] == true then
+        write_pref_attribute(device, "\x00\x02\x00\x01\x00\x00\x00")
+      else
+        write_pref_attribute(device, "\x00\x02\x00\x00\x00\x00\x00")
+      end
+
+      device.thread:call_with_delay(2, function(d)
+        read_pref_attribute(device)
+      end)
+    end
+
+    if device.preferences[softTouchPreferenceId] ~= args.old_st_store.preferences[softTouchPreferenceId] then
+      if device.preferences[softTouchPreferenceId] == true then
+        write_pref_attribute(device, "\x00\x08\x00\x00\x00\x00\x00")
+      else
+        write_pref_attribute(device, "\x00\x08\x00\x00\x00\x01\x00")
+      end
+    end
+  end
+
+  device:send(Groups.server.commands.RemoveAllGroups(device))
+end
+
+local function setInitializedState_handler(driver, device, command)
+  write_pref_attribute(device, "\x00\x01\x00\x00\x00\x00\x00")
+  device:emit_event(initializedstate.initializedState.initializing())
+
+  device.thread:call_with_delay(2, function(d)
+    local lastLevel = device:get_latest_state("main", capabilities.windowShadeLevel.ID, capabilities.windowShadeLevel.shadeLevel.NAME) or 0
+    if lastLevel > 0 then
+      device:set_field(INIT_STATE, INIT_STATE_CLOSE)
+      send_close_cmd(device, command.component)
+    else
+      device:set_field(INIT_STATE, INIT_STATE_OPEN)
+      send_open_cmd(device, command.component)
+    end
+  end)
+end
+
+local function emit_level_event(device, shadeLevel)
+  if shadeLevel == 100 then
+    device:emit_event(capabilities.windowShade.windowShade.open())
+  elseif shadeLevel == 0 then
+    device:emit_event(capabilities.windowShade.windowShade.closed())
+  else
+    device:emit_event(capabilities.windowShade.windowShade.partially_open())
+  end
+end
+
+local function window_shade_level_cmd(driver, device, command)
+  local level = command.args.shadeLevel
+  device:send_to_component(command.component, WindowCovering.server.commands.GoToLiftPercentage(device, level))
+  device:emit_event(capabilities.windowShadeLevel.shadeLevel(level))
+end
+
+local function window_shade_open_cmd(driver, device, command)
+  local shadeLevel = device:get_field(SHADE_LEVEL) or 0
+  send_open_cmd(device, command.component)
+  emit_level_event(device, shadeLevel)
+end
+
+local function window_shade_close_cmd(driver, device, command)
+  local shadeLevel = device:get_field(SHADE_LEVEL) or 0
+  send_close_cmd(device, command.component)
+  emit_level_event(device, shadeLevel)
+end
+
+local function window_shade_pause_cmd(driver, device, command)
+  local shadeLevel = device:get_field(SHADE_LEVEL) or 0
+  device:send_to_component(command.component, WindowCovering.server.commands.Stop(device))
+  emit_level_event(device, shadeLevel)
+end
+
+local function motion_state_attr_handler(driver, device, value, zb_rx)
+  local state = value.value
+  device:set_field(SHADE_STATE, state)
+
+  if state == 0 then

```

Comment:
Any documentation/comments on what these state values mean?

Reply:
In order to use the curtain controller, the curtain must be fully opened and completely closed by executing the 'initialization' function. After this process, the curtain controller remembers the length of the rail and can control percentage level. This is to remember the position state of the curtain.